### ISSUE TITLE
Direct connect Apple Watch

### ISSUE THREAD

**Issue by A-w-K** [JohanDegraeve/xdripswift#280]
Would it be possible to use the Apple Watch as primary device and collect data from the sensor?

**Comment by wiegerthefarmer** [comment]
I don't know if Apple has released that functionality in the watch. (the ability for the watch to connect to bluetooth devices). But I do think it's coming soon. There has been work in DiaBLE to have the watch connect directly to a libre patch that has bluetooth in it (Libre2,3,sense)

**Comment by gui-dos** [comment]
@A-w-K:

Yep! @JohanDegraeve confirmed me that you can reconnect to an European Libre 2 from a different device than the one you activated the BLE streaming from, since he could keep on working with his iPhone SE1 after sending the NFC command `1A 1E` on his wife’s phone.

@JtxGit also confirmed me that you can create an independent WearOS app after activating the Libre from a different Android device like I succeeded not only on watchOS but also from iPadOS and macOS (very probably it is even possible on tvOS…)

At each new connection you have to “login” to the `F001` characteristic writing an encoded "payload" providing three crucial parameters:

- a 32-bit “unlock code” established when sending `A1 1E`.The vendor uses the timestamp of the activation time but it has been discovered that any Int works (DiaBox and xDrip+ use the same, `42`)

- the sensor `patchInfo` at the time of the BLE stream activation (on a Libre 2 the last two bytes of six change continuously)

- an “unlock count” that has to be incremented by one each time.

`A1 1E` can be sent any number of times and the counter is simply reset to 0. That's the simple reason of the “broken connections” but there is no risk to waste a sensor like the FaceBook evangelists are saying… LibreLink itself, when reinstalled, can restore its own alarms.

**Comment by LiroyvH** [comment]
Interestingly, this is probably possible with Anubis G6 as well since it supports three channels instead of two. (Pump + iPhone + Apple Watch could all independently connect to the transmitter). And for users without pump or reader, it can *maybe* also be done with official G6 transmitters, but that may pose a challenge since those channels are specifically targeted at one reader device and one smartphone.

**Comment by tromanisagit** [comment]
I'm not sure how they do it but ambrosia linkblucon does this.  Ie scan on phone then move bt to watch.

**Comment by LiroyvH** [comment]
@tromanisagit 
They connect to their transmitter rather than the FSL. I've used their app in the past btw and it was extremely temperamental, hope we can do muuuuuch better if we implement something like this. ;)

**Comment by tromanisagit** [comment]
So - whats the best approach here?  create watch app? can that do all the nfc and other bits? or move the bluetooth to the watch after setup - like ambrosia?  not sure if the bluetooth can be moved? i assume amb are moving the phone side of the connection to the watch by moving the bluetooth mac?  i’ll have to read up on the bluetooth conneciton method.

**Comment by robster7674** [comment]
Looks like Apple has opened up the Watch for bluetooth devices:

Connect Bluetooth devices to Apple Watch
https://developer.apple.com/videos/play/wwdc2021/10005/

**Comment by miguelpruivo** [comment]
Bumping this thread. I've seen multiple devices connecting directly into the watch — for instance, my Nuki Smart Lock connects directly to the watch through BLE when no iPhone is near by.

This would be an amazing addition to xdrip as a fallback option (if iPhone is available, use the iPhone, else the watch) because for me, for example, everytime I go jogging, I have to take my iPhone with me or the libre reader just to have the readings whenever I need.

**Comment by LiroyvH** [comment]
@miguelpruivo pretty sure that's a special API for smarthome appliances, same as car keys. That's "cheating" a bit :P

**Comment by miguelpruivo** [comment]
@LiroyvH cant it be used for an app that’s not in the AppStore anyway? 😁

**Comment by LiroyvH** [comment]
@miguelpruivo Not sure! Apple has been rather protective of the Watch with entitlements, probably due to battery drain concerns. I'll have a look. I'm not sure if this can be made reliable anyway with how very conservative the Watch is due to battery constraints. For good reason, but still annoying heh.

**Comment by tromanisagit** [comment]
Ambrosia have just announced a similar feature, but its not with the direct to libre.

**Comment by gui-dos** [comment]
I own an Apple Watch 4 and I am not working actively on my [DiaBLE](https://github.com/gui-dos/DiaBLE/tree/main/DiaBLE%20Watch%20Extension). Notice however that the Series 6/8 have enhanced capabilities to keep the BLE scans and connections in background: https://developer.apple.com/videos/play/wwdc2022/10135/.

SwiftUI was started by the Apple Watch team and you can create independent apps which can share most of the code base: the Core Bluetooth framework is the same.

I am wearing a Libre 1 now but I can connect directly to a Libre 2 / 3 and decrypt the former data by copying&pasting the login parameters (unlock code and count) from the iPhone used to activate the BLE streaming.

As a workaround the get the encrypted Libre 3 data you can scrape directly LibreLinkUp by using 
your LibreView credentials. This is the technique used currently by Glucose Direct by @creepymonster but it works with all sensors, of course:
![incoming-DD1D8292-1520-4EAC-9B33-D0164FA540E5](https://user-images.githubusercontent.com/7220550/211783490-4fa407a1-474c-40f8-b6ce-2c8fcd0da2a1.PNG)

**Comment by LiroyvH** [comment]
@tromanisagit Ambrosia has had it for a few years, but it worked exceptionally unreliable when I tried it. Maybe they fixed it now, but wouldn't get hopes up.

**Comment by gui-dos** [comment]
A direct connection with a Bubble is also possiBLE but I got no feedback from the Bubble/MiaoMiao teams. I never managed the background connections properly but I thought my screenshots would impress them anyway...


![incoming-F305BD54-84DA-447C-9CE9-84FCA64EA5EC](https://user-images.githubusercontent.com/7220550/212666695-7c37b068-793b-4904-9912-ee91aa6b609c.PNG)

**Comment by tromanisagit** [comment]
Just a note.  Its interesting that dexcon G7 now has direct to watch.  Looks like its a start on phone and shift to watch method.  Interesting to see how they do that.

**Comment by LiroyvH** [comment]
@tromanisagit There's an extra slot on it. The Dexcom transmitter always controls the state, this has been the same for the past generations of Dexcom's. So essentially: the only reason you have to start it on the phone, is because the Watch doesn't have an option built-in to start it. For the rest it works exactly the same as other devices. :)

**Comment by ricardosllm** [comment]
Bumping this thread. Long term developer, libre and xdrip4ios user (I guess over 10 years now with the libre 1,2) and really interested in getting a direct connection working, more than happy to work on it, would be great if any research into this has already been done to not start from scratch.

Overall a direct connection to the watch would make a lot of sense as we're more often with our watch than with our phone. 

Feel free to ping me to work on this.

**Comment by m-a-v** [comment]
@ricardosllm Eventually @gui-dos can give you more insights what is already possible with his prototype or you just test it ;-).

https://github.com/gui-dos/DiaBLE

**Comment by rnederstigt** [comment]
For the past few days I've been bullying AI to code a patch allowing the app to connect a FSL2 directly to an Apple Watch and back. 

I'd like to stress that this is currently a foreground proof of concept. Reliable continuous background collection remains unresolved, and it does not add standalone Watch alarms or automatic failover. NFC reclaim changes sensor credentials and can disrupt other apps’ connections or alarms. And of course the big caveat, coded by AI and limited testing performed. 

An intended use case would be open water sports when carrying a phone is cumbersome, and continuous BLE connections would be intermittent at best, yet one would still be able to monitor glucose values when necessary, i.e essentially functioning as a FGM.

https://github.com/rnederstigt/xdripswift/tree/experimental/libre2watch
